### PR TITLE
Expose window functions in the Python API

### PR DESCRIPTION
**Pull request by jarandaf**

### CODE REVIEW

File: tools/pythonpkg/src/pyrelation.cpp [hunk 1]
```diff
@@ -252,27 +252,35 @@ string DuckDBPyRelation::ToSQL() {
...
 		expr = projected_columns + ", ";
 	}
-	for (idx_t i = 0; i < input.size(); i++) {
-		if (function_parameter.empty()) {
-			expr += function_name + "(" + input[i] + ") " + window_function;
-		} else {
-			expr += function_name + "(" + input[i] + "," + function_parameter + ")" + window_function;
-		}
+	if (input.size() == 0 && !function_parameter.empty()) {
+		expr += function_name + "(" + function_parameter + ((ignore_nulls) ? " ignore nulls) " : ") ") + window_spec;
```
Comment (Tishj) [inline tools/pythonpkg/src/pyrelation.cpp:271]: I would prefer to early-return here so we don't need the indented `else` block

File: tools/pythonpkg/src/pyrelation.cpp [hunk 2]
```diff
@@ -401,6 +409,100 @@ unique_ptr<DuckDBPyRelation> DuckDBPyRelation::CumMin(const string &aggr_columns
 	return GenericWindowFunction("min", aggr_columns);
 }
 
+unique_ptr<DuckDBPyRelation> DuckDBPyRelation::RowNumber(const string &window_spec, const string &projected_columns) {
+	auto expr = GenerateExpressionList("row_number", "*", "", "", false, projected_columns, window_spec);
+	return make_uniq<DuckDBPyRelation>(rel->Project(expr));
```
Comment (Tishj) [inline tools/pythonpkg/src/pyrelation.cpp:414]: Slight nitpick but is there a reason we don't use GenericWindowFunction in these functions?
Currently these methods are only 2 lines long but if this scales in the future we would be duplicating these changes across all these functions, so I'd prefer to turn these into just a call to GenericWindowFunction if possible

File: tools/pythonpkg/src/pyrelation.cpp [hunk 3]
```diff
@@ -401,6 +409,100 @@ unique_ptr<DuckDBPyRelation> DuckDBPyRelation::CumMin(const string &aggr_columns
...
+unique_ptr<DuckDBPyRelation> DuckDBPyRelation::FirstValue(const string &column, const string &window_spec,
+                                                          const bool &ignore_nulls, const string &projected_columns) {
+	auto expr = GenerateExpressionList("first_value", column, "", "", ignore_nulls, projected_columns, window_spec);
+	return make_uniq<DuckDBPyRelation>(rel->Project(expr));
+}
+
+unique_ptr<DuckDBPyRelation> DuckDBPyRelation::First(const string &column, const string &window_spec,
+                                                     const bool &ignore_nulls, const string &projected_columns) {
+	// alias for first_value
+	return FirstValue(column, window_spec, ignore_nulls, projected_columns);
```
Comment (Tishj) [inline tools/pythonpkg/src/pyrelation.cpp:482]: same as before, please use DefineMethod in `pyrelation/initialize.cpp` instead

File: tools/pythonpkg/src/pyrelation.cpp [hunk 4]
```diff
@@ -401,6 +409,100 @@ unique_ptr<DuckDBPyRelation> DuckDBPyRelation::CumMin(const string &aggr_columns
...
+unique_ptr<DuckDBPyRelation> DuckDBPyRelation::LastValue(const string &column, const string &window_spec,
+                                                         const bool &ignore_nulls, const string &projected_columns) {
+	auto expr = GenerateExpressionList("last_value", column, "", "", ignore_nulls, projected_columns, window_spec);
+	return make_uniq<DuckDBPyRelation>(rel->Project(expr));
+}
+
+unique_ptr<DuckDBPyRelation> DuckDBPyRelation::Last(const string &column, const string &window_spec,
+                                                    const bool &ignore_nulls, const string &projected_columns) {
+	// alias for last_value
+	return LastValue(column, window_spec, ignore_nulls, projected_columns);
```
Comment (Tishj) [inline tools/pythonpkg/src/pyrelation.cpp:494]: Also DefineMethod

**Comment by Tishj** [comment]
Thanks for the PR, maybe we can add a `window` method to use any window function, just as we have the `apply` method to aggregate using any aggregate function.

You will also need to update the `duckdb-stubs/__init__.py` file with the newly added methods

File: tools/pythonpkg/src/pyrelation.cpp (hunk 2 above)
Comment (jarandaf) [reply tools/pythonpkg/src/pyrelation.cpp:414]: Fair point. I will try to refactor this.

**Review by Tishj — APPROVED** [review]
Thanks for the changes, I think this is ready if CI passes 👍

**Comment by jarandaf** [comment]
Aggregation functions within a windowing context are still not available. I plan to support them in the next days.

**Comment by jarandaf** [comment]
The PR should be now ready for review @pdet @Tishj.

**Review by Tishj — APPROVED** [review]
Thanks for the changes!

**Pull request merged**